### PR TITLE
tagging

### PR DESCRIPTION
**Pull request by clarali210**

### CODE REVIEW

**Comment by clarali210** [comment]
Seems to work now, need to try it out in full framework to see how well it merges with gold sets evaluation and exports etc.

File: config.py [hunk 1]
```diff
@@ -94,7 +94,10 @@ class Config(object):
         'use_top_tfidf_sampling': False,
         'learn_to_select':True,
         'num_samples': 1,
-        'debug_mode': False
+        'debug_mode': False,
+        'vocab_size': 1e6,
```
[the remark this answers is not in the document]
Comment (clarali210) [reply config.py]: should i change the number?

File: oncotext/datasets/patholgy_dataset.py [hunk 2]
```diff
@@ -17,12 +17,10 @@
 
 class PathologyDataset(data.Dataset):
```
[the remark this answers is not in the document]
Comment (clarali210) [reply oncotext/datasets/patholgy_dataset.py]: resolved.

File: oncotext/utils/dataset_factory.py [hunk 3]
```diff
@@ -11,29 +12,43 @@ def get_oncotext_dataset_train(all_reports, label_maps, args, text_key, vocab_si
...
+                                                'dev')
     return train_data, dev_data
 
 
-
-
-def get_oncotext_dataset_test(reports, label_maps, args, text_key, vocab_size):
-    test_data = oncotext.datasets.patholgy_dataset.PathologyDataset(args,reports, label_maps, text_key, 'test', vocab_size=vocab_size)
+def get_oncotext_dataset_test(reports, label_maps, args, text_key):
+    if label_maps[args.aspect][0] == "NUM":
```
[the remark this answers is not in the document]
Comment (clarali210) [reply oncotext/utils/dataset_factory.py]: resolved.

File: oncotext/datasets/patholgy_dataset_tagging.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    def processLine(self, sample):
+        text = sample[self.text_key].split()[:self.args.max_length]
+        text_indx = [self.hash(token) for token in text]
+        if len(text_indx) < self.args.max_length:
+            nil_indx = 0
+            text_indx.extend([nil_indx for _ in range(self.args.max_length - len(text_indx))])
+
+        x = torch.LongTensor([text_indx])
+
+        label_indx = [0 for _ in range(len(text_indx))]
```
[the remark this answers is not in the document]
Comment (clarali210) [reply oncotext/datasets/patholgy_dataset_tagging.py]: was trying to keep consistent with text_indx, its not exactly the original label since its been converted to array of 0s and 1s

File: config.py [hunk 5]
```diff
@@ -59,7 +59,7 @@ class Config(object):
     RATIONALE_NET_CONFIG = {
         'cuda': True,
         'num_workers': 8,
-        'batch_size': 128,
+        'batch_size': 32, #128,
```
[the remark this answers is not in the document]
Comment (clarali210) [reply config.py]: resolved.

File: oncotext/datasets/patholgy_dataset_tagging.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        self.dataset = reports
+        self.text_key = text_key
+        self.samples = copy.deepcopy( self.dataset)
+        self.class_balance = {'0': 0, '1': 0}
+        for i, sample in tqdm.tqdm(enumerate(self.samples)):
+            sample['x'], sample['y'], sample['val'], sample['match'] = self.processLine(sample)
+            sample['i'] = i
+
+            self.class_balance[sample['match']] += 1
+
```
[the remark this answers is not in the document]
Comment (clarali210) [reply oncotext/datasets/patholgy_dataset_tagging.py]: yes, added inline comments for explanation.

File: oncotext/datasets/patholgy_dataset_tagging.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        print ("Class balance", self.class_balance)
+        args.num_class = args.num_tags
+
+                        
+    def hash(self, token):
+        """Unsigned 32 bit murmurhash for feature hashing."""
+        return murmurhash3_32(token, positive=True) % self.args.vocab_size
+
+    
+    ## Convert one line from beer dataset to {Text, Tensor, Labels}
```
[the remark this answers is not in the document]
Comment (clarali210) [reply oncotext/datasets/patholgy_dataset_tagging.py]: resolved.

File: oncotext/datasets/patholgy_dataset_tagging.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        if len(text_indx) < self.args.max_length:
+            nil_indx = 0
+            text_indx.extend([nil_indx for _ in range(self.args.max_length - len(text_indx))])
+
+        x = torch.LongTensor([text_indx])
+
+        label_indx = [0 for _ in range(len(text_indx))]
+        val = 'NA'
+        match = '1'
+        if self.name != "test":
```
[the remark this answers is not in the document]
Comment (clarali210) [reply oncotext/datasets/patholgy_dataset_tagging.py]: i was trying to get it to work for labels that weren't consecutive in the report text, realized wasn't necessary and is simplified now.

File: config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yala) [reply config.py]: nah, it's fine. I remember now, it was to reduce the probablity of hash collision or someting

File: oncotext/datasets/pathology_dataset_classifier.py [hunk 9]
```diff
@@ -15,14 +15,12 @@
 random.seed(0)
 
 
-class PathologyDataset(data.Dataset):
+class PathologyDatasetClassifier(data.Dataset):
```
Comment (yala) [inline oncotext/datasets/pathology_dataset_classifier.py]: Would rename to "PathologyClassificationDataset" 
It's primarily a dataset, not a classifer

File: oncotext/datasets/pathology_dataset_tagger.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+import rationale_net.utils.dataset as utils
+import random
+import copy
+from sklearn.utils import murmurhash3_32
+
+SMALL_TRAIN_SIZE = 800
+random.seed(0)
+
+
+class PathologyDatasetTagger(data.Dataset):
```
Comment (yala) [inline oncotext/datasets/pathology_dataset_tagger.py]: see above

File: oncotext/rationale_net_wrapper.py [hunk 11]
```diff
@@ -99,12 +120,36 @@ def label_reports(name, un_reports, config, logger):
...
+            
         except Exception as e:
             logger.warn("RN Wrapper. {} model failed to label reports! Following Exception({}). Populating all reports with 0 label".format(diagnosis, e))
             preds = np.zeros(len(test_data), dtype=int)
 
-        for i in range(len(test_data)):
-            prediction = label_maps[diagnosis][preds[i]]
-            test_data.dataset[i][diagnosis] = prediction
+        if label_maps[diagnosis][0] == "NUM":
+            try:
```
Comment (yala) [inline oncotext/rationale_net_wrapper.py]: I would recomment refactor this as:
```
for i in range(len(data))):
       text = blah
       pred = helper.get_pred(blah)
```
and moving the details to a helper function/file

File: config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (clarali210) [reply config.py]: resolved.

File: oncotext/datasets/pathology_dataset_classifier.py (hunk 9 above)
Comment (clarali210) [reply oncotext/datasets/pathology_dataset_classifier.py]: resolved

File: oncotext/datasets/pathology_dataset_tagger.py (hunk 10 above)
Comment (clarali210) [reply oncotext/datasets/pathology_dataset_tagger.py]: resolved.

File: oncotext/utils/dataset_factory.py [hunk 12]
```diff
@@ -11,30 +13,58 @@ def get_oncotext_dataset_train(all_reports, label_maps, args, text_key, vocab_si
...
     return test_data
 
+
 def get_embedding_tensor(config, args):
     embeddings =  pickle.load(open(config['EMBEDDING_PATH'],'rb'))
     args.embedding_dim = embeddings.shape[1]
     return embeddings
+
+
+def predsToLabels(preds, test_data, label_maps, diagnosis, args, text_key, logger):
```
Comment (yala) [inline oncotext/utils/dataset_factory.py]: nitpicky but 

all our function names are "blah_blah", so please change this to that convention instead of Camel case.

File: oncotext/utils/dataset_factory.py [hunk 13]
```diff
@@ -11,30 +13,58 @@ def get_oncotext_dataset_train(all_reports, label_maps, args, text_key, vocab_si
...
+            preds = np.reshape(preds, (len(test_data), args.max_length))
+        except Exception as e:
+            logger.warn("RN Wrapper. {} model returned incorrectly sized labels {}! Following Exception({}). Populating all reports with 0 label".format(diagnosis, (len(preds), len(test_data)), e))
+            preds = np.zeros((len(test_data), args.max_length), dtype=int)
+
+        for i in range(len(test_data)):
+            if 1 in preds[i]:
+                text = test_data.dataset[i][text_key].split()
+                inds = np.where(preds[i]== 1)[0]
+                if len(inds) > 0:
```
Comment (yala) [inline oncotext/utils/dataset_factory.py]: why would this ever be false?

Comment (yala) [reply oncotext/utils/dataset_factory.py]: I think this function should be called "get_labels_from_predictions".

And you should use two helper functions:
"get_labels_from_tagging_predictions" and "get_labels_from_classification_predicitons"
